File: Examples/test-suite/php/director_classes_runme.php
<?php
require "tests.php";

// No new functions
check::functions(array());
// New classes
check::classes(array('director_classes', 'Base', 'BaseClass', 'Caller', 'Derived', 'DerivedClass', 'DoubleHolder'));
// New vars
check::globals(array('PrintDebug'));

class PHPDerived extends Base {
  function Val(DoubleHolder $x) { return $x; }
  function Ref(DoubleHolder $x) { return $x; }
  function Ptr(?DoubleHolder $x) { return $x; }
  function ConstPtr(?DoubleHolder $x) { return $x; }
  function ConstPtrRef(?DoubleHolder $x) { return $x; }
  function FullyOverloaded($x) {
    return preg_replace('/Base/', 'PHPDerived', parent::FullyOverloaded($x));
  }
  function SemiOverloaded($x) {
    # PHP use one callback for both types!
    if (is_int($x)) {
        # PHPDerived::SemiOverloaded(int) overload
        return preg_replace('/Base/', 'PHPDerived', parent::SemiOverloaded($x));
    }
    # Call Base::SemiOverloaded(bool)
    return parent::SemiOverloaded($x);
  }
  function DefaultParms(int $x, float $y = 1.125) { /* Same default as C++! */
    return preg_replace('/Base/', 'PHPDerived', parent::DefaultParms($x, $y));
  }
}


function makeCalls($caller, $base, $TODO) {
  $bname = get_class($base);
  $caller->set($base);
  $dh = new DoubleHolder(444.555);
  check::equal($caller->ValCall($dh)->val, $dh->val, "$bname.Val");
  check::equal($caller->RefCall($dh)->val, $dh->val, "$bname.Ref");
  check::equal($caller->PtrCall($dh)->val, $dh->val, "$bname.Ptr");
  check::equal($caller->ConstPtrCall($dh)->val, $dh->val, "$bname.ConstPtr");
  if (!$TODO) {
  # TODO Need fixing!
  # Use SWIGTYPE_p_p_DoubleHolder instead of SWIGTYPE_p_DoubleHolder!
  check::equal($caller->ConstPtrRefCall($dh)->val, $dh->val, "$bname.ConstPtrRef");
  }
  check::equal($caller->FullyOverloadedCall(1),
	       "{$bname}::FullyOverloaded(int)",
	       "$bname.FullyOverloaded(int)");
  check::equal($caller->FullyOverloadedCall(false),
	       "{$bname}::FullyOverloaded(bool)",
	       "$bname.FullyOverloaded(bool)");
  check::equal($caller->SemiOverloadedCall(-678),
	       "{$bname}::SemiOverloaded(int)",
	       "$bname.SemiOverloaded(int)");
  check::equal($caller->SemiOverloadedCall(false),
	       "Base::SemiOverloaded(bool)",
	       "$bname.SemiOverloaded(bool)");
  check::equal($caller->DefaultParmsCall(10, 2.2),
	       "{$bname}::DefaultParms(int, double)",
	       "$bname.DefaultParms(int, double)");
  check::equal($caller->DefaultParmsCall(10),
	       "{$bname}::DefaultParms(int)",
	       "$bname.DefaultParms(int)");
  $caller->reset();
}

$c = new Caller();
makeCalls($c, new Base(100.0), false);
makeCalls($c, new Derived(200.0), false);
makeCalls($c, new PHPDerived(300.0), true);

check::done();

File: Examples/test-suite/perl5/director_classes_runme.pl
use strict;
use warnings;
use Test::More tests => 35;
BEGIN { use_ok 'director_classes' }
require_ok 'director_classes';

{
  package PerlDerived;
  use base 'director_classes::Base';
  sub Val { $_[1] }
  sub Ref { $_[1] }
  sub Ptr { $_[1] }
  sub ConstPtr { $_[1] }
  sub ConstPtrRef { $_[1] }
  sub FullyOverloaded {
    my $rv = shift->SUPER::FullyOverloaded(@_);
    $rv =~ s/Base/__PACKAGE__/sge;
    return $rv;
  }
  sub SemiOverloaded {
    my $self = shift;
    my $val = shift;
    # Perl use one callback for both SemiOverloaded() methods.
    # Usually we use the values type and number of values to distinguish.
    # But as Perl tread boolean as number (as old C does),
    # We fake it by detecting the value itself.
    my $rv = $self->SUPER::SemiOverloaded($val);
    if($val eq "") { # detect SemiOverloaded(bool)
        # Convert SemiOverloaded(int) to SemiOverloaded(bool)
        $rv =~ s/int/bool/;
    } else { # detect SemiOverloaded(int)
        # Convert Base::SemiOverloaded(int) to PerlDerived::SemiOverloaded(int)
        $rv =~ s/Base/__PACKAGE__/se;
    }
    return $rv;
  }
  sub DefaultParms {
    my $rv = shift->SUPER::DefaultParms(@_);
    $rv =~ s/Base/__PACKAGE__/sge;
    return $rv;
  }
}

{
  my $c = director_classes::Caller->new();
  makeCalls($c, director_classes::Base->new(100.0));
  makeCalls($c, director_classes::Derived->new(200.0));
  makeCalls($c, PerlDerived->new(300.0));
}

sub makeCalls { my($caller, $base) = @_;
  my $bname = ref $base;
  $bname = $1 if $bname =~ /^director_classes::(.*)$/;
  $caller->set($base);
  my $dh = director_classes::DoubleHolder->new(444.555);
  is($caller->ValCall($dh)->{val}, $dh->{val}, "$bname.Val");
  is($caller->RefCall($dh)->{val}, $dh->{val}, "$bname.Ref");
  is($caller->PtrCall($dh)->{val}, $dh->{val}, "$bname.Ptr");
  is($caller->ConstPtrCall($dh)->{val}, $dh->{val}, "$bname.ConstPtr");
  is($caller->ConstPtrRefCall($dh)->{val}, $dh->{val}, "$bname.ConstPtrRef");
  is($caller->FullyOverloadedCall(1),
      "${bname}::FullyOverloaded(int)",
      "$bname.FullyOverloaded(int)");
  is($caller->FullyOverloadedCall(''),
      "${bname}::FullyOverloaded(bool)",
      "$bname.FullyOverloaded(bool)");
  is($caller->SemiOverloadedCall(-678),
      "${bname}::SemiOverloaded(int)",
      "$bname.SemiOverloaded(int)");
  is($caller->SemiOverloadedCall(''),
      "Base::SemiOverloaded(bool)",
      "$bname.SemiOverloaded(bool)");
  is($caller->DefaultParmsCall(10, 2.2),
      "${bname}::DefaultParms(int, double)",
      "$bname.DefaultParms(int, double)");
  is($caller->DefaultParmsCall(10),
      "${bname}::DefaultParms(int)",
      "$bname.DefaultParms(int)");
  $caller->reset();
}
